crypto/bls_core.c

```diff
@@ -1,541 +1,503 @@
+  E2_copy(&elemsG2[0], BLS12_381_minus_g2);
+
+  // map all hashes to G1
+  int offset = 0;
+  for (int i = 1; i < nb_hashes + 1; i++) {
+    // elemsG1[i] = h
+    // hash to G1
+    map_to_G1(&elemsG1[i], &hashes[offset], len_hashes[i - 1]);
+    offset += len_hashes[i - 1];
+  }
+
+  // aggregate public keys mapping to the same hash
+  offset = 0;
+  for (int i = 1; i < nb_hashes + 1; i++) {
+    // elemsG2[i] = agg_pk[i]
+    E2_sum_vector(&elemsG2[i], &pks[offset], pks_per_hash[i - 1]);
+    offset += pks_per_hash[i - 1];
+  }
+
+  // multi pairing
+  Fp12 e;
+  Fp12_multi_pairing(&e, elemsG1, elemsG2, nb_hashes + 1);
+  if (Fp12_is_one(&e)) {
+    ret = VALID;
+  } else {
+    ret = INVALID;
+  }
 
 out:
-    for (int i=0; i < nb_hashes+1; i++) {
-        ep_free(elemsG1[i]);
-        ep2_free(elemsG2[i]);
-    }
-    free(elemsG2);
+  free(elemsG2);
 outG2:
-    free(elemsG1);
+  free(elemsG1);
 outG1:
-    return ret;
+  return ret;
 }
 
-
-// Verifies the validity of an aggregated BLS signature under distinct public keys.
+// Verifies the validity of an aggregated BLS signature under distinct public
+// keys.
 //
-// Each key is mapped to a set of messages, so that the verification equation is 
-// optimized to compute one pairing per public key. 
+// Each key is mapped to a set of messages, so that the verification equation is
+// optimized to compute one pairing per public key.
 // - nb_pks is the number of the public keys in the map.
 // - pks is pointer to all pks in order where the key at index i
-//   is mapped to hashes_per_pk[i] hashes. 
+//   is mapped to hashes_per_pk[i] hashes.
 // - the messages (hashes) are flattened in hashes in the same public key order,
 //  each with a length in len_hashes.
 //
 // membership check of the signature in G1 is verified in this function
 // membership check of pks in G2 is not verified in this function
-// the membership check is separated to allow optimizing multiple verifications using the same pks
-int bls_verifyPerDistinctKey(const byte* sig, 
-                         const int nb_pks, const ep2_st* pks, const uint32_t* hashes_per_pk,
-                         const byte* hashes, const uint32_t* len_hashes){
-
-    int ret = UNDEFINED; // return value
-    
-    ep_t* elemsG1 = (ep_t*)malloc((nb_pks + 1) * sizeof(ep_t));
-    if (!elemsG1) goto outG1;
-    ep2_t* elemsG2 = (ep2_t*)malloc((nb_pks + 1) * sizeof(ep2_t));
-    if (!elemsG2) goto outG2;
-    for (int i=0; i < nb_pks+1; i++) {
-        ep_new(elemsG1[i]);
-        ep2_new(elemsG2[i]);
+// the membership check is separated to allow optimizing multiple verifications
+// using the same pks
+int bls_verifyPerDistinctKey(const byte *sig, const int nb_pks, const E2 *pks,
+                             const uint32_t *hashes_per_pk, const byte *hashes,
+                             const uint32_t *len_hashes) {
+
+  int ret = UNDEFINED; // return value
+
+  E1 *elemsG1 = (E1 *)malloc((nb_pks + 1) * sizeof(E1));
```

Comment:
have you considered using `calloc` for allocations of arrays?

Reply:
Not really. The initialization of memory isn't needed, but are you referring to possible non-contiguous memory issues?